src/openstack_cloud/__init__.py

```diff
@@ -19,31 +18,15 @@
 CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
 
 
-# Make sure we can import openstack, if not remove the old openstacksdk library and retry.
 # This is a workaround for https://bugs.launchpad.net/juju/+bug/2058335
-def _remove_old_openstacksdk_lib() -> None:  # pragma: no cover
-    """Remove the old openstacksdk library if it exists."""
-    try:
-        unit_name = os.environ["JUJU_UNIT_NAME"].replace("/", "-")
-        venv_dir = Path(f"/var/lib/juju/agents/unit-{unit_name}/charm/venv/")
-        openstacksdk_dirs = list(venv_dir.glob("openstacksdk-*.dist-info"))
-        # use error log level as logging may not be fully initialized yet
-        logger.error("Found following openstack dirs: %s", openstacksdk_dirs)
-        if len(openstacksdk_dirs) > 1:
-            openstacksdk_dirs.sort()
-            for openstacksdk_dir in openstacksdk_dirs[:-1]:
-                logger.error("Removing old openstacksdk library: %s", openstacksdk_dir)
-                shutil.rmtree(openstacksdk_dir)
-        else:
-            logger.error(
-                "No old openstacksdk library to remove. "
-                "Please reach out to the charm dev team for further advice."
-            )
-    except OSError:
-        logger.exception(
-            "Failed to remove old openstacksdk library. "
-            "Please reach out to the charm dev team for further advice."
-        )
+def _remove_residual_venv_dirs() -> None:  # pragma: no cover
+    """Remove the residual empty directories from last revision if it exists."""
+    unit_name = os.environ["JUJU_UNIT_NAME"].replace("/", "-")
+    venv_dir = Path(f"/var/lib/juju/agents/unit-{unit_name}/charm/venv/")
+    for path in venv_dir.iterdir():
+        if path.is_dir() and not os.listdir(path):
+            logger.warning("Removing residual empty dir: %s", path)
+            path.rmdir()
```

Comment:
The code is no longer specifically targeting an openstack lib. Can you move this code somewhere else and make sure it is called at the top of the charm.py module? Otherwise it looks a bit odd to a reader who does not have the internal knowledge of why we are doing this.

Reply:
This operation has a performance penalty as it walks the entire `venv` directory, which can be quite large. Therefore, I think it might be best to call this function only when necessary.

Reply:
Okay, I've tested this in the real environment, and the entire process takes less than a second to finish, so I guess it's okay to run this every time a hook is fired. But please make sure it's removed once this problem is fixed in Juju. Thanks!